src/BoltOn.Data.EF/Repository.cs

```diff
@@ -75,6 +74,22 @@ public virtual async Task DeleteAsync(TEntity entity, object options = null, Can
 			await SaveChangesAsync(entity, cancellationToken);
 		}
 
+		public async Task<IEnumerable<TEntity>> AddAsync(IEnumerable<TEntity> entities, object options = null, CancellationToken cancellationToken = default)
+		{
+			foreach (var entity in entities)
+			{
+				DbSets.Add(entity);
+			}
+
+			foreach (var entity in entities)
+			{
+				PublishEvents(entity);
+			}
+			await DbContext.SaveChangesAsync(cancellationToken);
+
+			return entities;
+		}
+
 		protected async Task SaveChangesAsync(TEntity entity, CancellationToken cancellationToken = default)
```

Comment:
what if create another SaveChangesAsync method with IEnumerable < TEntity > and CancellationToken as params? and call it from the new AddAsync method?

the reason why i abstracted SaveChangesAsync and PublishEvents is for those who override the base repository can call SaveChangesAsync method and which will take care publishing events... ideally PublishEvents could be private. make all the SaveChanges and PublishEvents virtual.

Reply:
@gokulm If PublishEvents method is marked as virtual, then why making it private, it could remain protected.

Reply:
@cthumburu do u see the need for making it protected virtual? if not, private should be good.

Reply:
@gokulm I thought PublishEvents was abstracted initially similar to SaveChangesAsync method for those who overrides the base repository and use the methods if needed. For SaveChangesAsync method, it make sense, I don't think anyone would use PublishEvents alone without persisting the changes. If that make sense, then I'll mark the method as private.